flight/PiOS/Common/pios_mpu.c

```diff
@@ -0,0 +1,1090 @@
+/**
+ ******************************************************************************
+ * @file       pios_mpu.c
+ * @author     dRonin, http://dRonin.org/, Copyright (C) 2016
+ * @author     Tau Labs, http://taulabs.org, Copyright (C) 2013
+ * @addtogroup PIOS PIOS Core hardware abstraction layer
+ * @{
+ * @addtogroup PIOS_MPU Invensense MPU Functions
+ * @{
+ * @brief Common driver for Invensense MPU-xxx0 6/9-axis IMUs
+ *****************************************************************************/
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ *
+ * Additional note on redistribution: The copyright and license notices above
+ * must be maintained in each individual source file that is a derivative work
+ * of this source file; otherwise redistribution is prohibited.
+ */
+
+#include "openpilot.h"
+#include "pios.h"
+
+#if defined(PIOS_INCLUDE_MPU)
+
+#include "pios_semaphore.h"
+#include "pios_thread.h"
+#include "pios_queue.h"
+#include "physical_constants.h"
+#include "taskmonitor.h"
+
+#include "pios_mpu_priv.h"
+
+#define PIOS_MPU_TASK_PRIORITY   PIOS_THREAD_PRIO_HIGHEST
+#ifdef PIOS_INCLUDE_MPU_MAG
+#define PIOS_MPU_TASK_STACK      600
+#else
+#define PIOS_MPU_TASK_STACK      512
+#endif // PIOS_INCLUDE_MPU_MAG
+
+#define PIOS_MPU_QUEUE_LEN       2
+
+#ifndef PIOS_MPU_SPI_HIGH_SPEED
+#define PIOS_MPU_SPI_HIGH_SPEED              20000000	// should result in 10.5MHz clock on F4 targets like Sparky2
+#endif // PIOS_MPU_SPI_HIGH_SPEED
+#define PIOS_MPU_SPI_LOW_SPEED               300000
+
+
+/**
+ * WHOAMI ids of each device, must be same length as pios_mpu_type
+ */
+static const uint8_t pios_mpu_whoami[PIOS_MPU_NUM] = {
+	[PIOS_MPU60X0] = 0x68,
+	[PIOS_MPU6500] = 0x70,
+	[PIOS_MPU9150] = 0x68,
+	[PIOS_MPU9250] = 0x71
+};
+
+/**
+ * I2C addresses to probe for device
+ */
+static const uint8_t pios_mpu_i2c_addr[] = {
+	0x68,
```

Comment:
What do you think about the approach here, https://github.com/TauLabs/TauLabs/pull/2206/files#diff-db1c8b7b78676a2aada18cdf6f5e4428R62, where the enum spells out the magic number?